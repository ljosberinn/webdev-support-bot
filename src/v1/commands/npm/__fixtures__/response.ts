export const response = [
  {
    date: '2020-03-19T19:53:13.309Z',
    description: 'React is a JavaScript library for building user interfaces.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
    ],
    name: 'react',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Jed Watson',
    },
    date: '2020-03-22T04:02:30.633Z',
    description: 'A Select control built with and for ReactJS',
    keywords: [
      'combobox',
      'form',
      'input',
      'multiselect',
      'react',
      'react-component',
      'select',
      'ui',
    ],
    links: {
      bugs: 'https://github.com/JedWatson/react-select/issues',
      homepage: 'https://github.com/JedWatson/react-select#readme',
      npm: 'https://www.npmjs.com/package/react-select',
      repository: 'https://github.com/JedWatson/react-select',
    },
    maintainers: [
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'cc.lee@live.com.au',
        username: 'gwyneplaine',
      },
      {
        email: 'jed.watson@me.com',
        username: 'jedwatson',
      },
      {
        email: 'mitchell@mitchellhamilton.me',
        username: 'mitchellhamilton',
      },
      {
        email: 'automation+react-select@thinkmill.com.au',
        username: 'react-select-release-bot',
      },
    ],
    name: 'react-select',
    publisher: {
      email: 'automation+react-select@thinkmill.com.au',
      username: 'react-select-release-bot',
    },
    scope: 'unscoped',
    version: '3.1.0',
  },
  {
    author: {
      name: 'Matt Zabriskie',
    },
    date: '2020-05-14T14:38:16.281Z',
    description: 'React draggable component',
    keywords: ['react', 'draggable', 'react-component'],
    links: {
      bugs: 'https://github.com/mzabriskie/react-draggable/issues',
      homepage: 'https://github.com/mzabriskie/react-draggable',
      npm: 'https://www.npmjs.com/package/react-draggable',
      repository: 'https://github.com/mzabriskie/react-draggable',
    },
    maintainers: [
      {
        email: 'mzabriskie@gmail.com',
        username: 'mzabriskie',
      },
      {
        email: 'samuel.trace.reed@gmail.com',
        username: 'strml',
      },
    ],
    name: 'react-draggable',
    publisher: {
      email: 'samuel.trace.reed@gmail.com',
      username: 'strml',
    },
    scope: 'unscoped',
    version: '4.4.2',
  },
  {
    author: {
      email: 'travis@souporserious.com',
      name: 'Travis Arnold',
      url: 'http://souporserious.com',
    },
    date: '2020-04-28T09:55:08.691Z',
    description: 'Official library to use Popper on React projects',
    keywords: [
      'react',
      'react-popper',
      'popperjs',
      'component',
      'drop',
      'tooltip',
      'popover',
    ],
    links: {
      bugs: 'https://github.com/popperjs/react-popper/issues',
      homepage: 'https://github.com/popperjs/react-popper',
      npm: 'https://www.npmjs.com/package/react-popper',
      repository: 'https://github.com/popperjs/react-popper',
    },
    maintainers: [
      {
        email: 'federico.zivolo@gmail.com',
        username: 'fezvrasta',
      },
      {
        email: 'ftntravis@gmail.com',
        username: 'souporserious',
      },
    ],
    name: 'react-popper',
    publisher: {
      email: 'federico.zivolo@gmail.com',
      username: 'fezvrasta',
    },
    scope: 'unscoped',
    version: '2.2.3',
  },
  {
    author: {
      name: 'Param Aggarwal',
    },
    date: '2020-04-26T08:29:08.419Z',
    description: 'Simple HTML5 drag-drop zone with React.js',
    keywords: ['react-component', 'react', 'drag', 'drop', 'upload'],
    links: {
      bugs: 'https://github.com/react-dropzone/react-dropzone/issues',
      homepage: 'https://github.com/react-dropzone/react-dropzone',
      npm: 'https://www.npmjs.com/package/react-dropzone',
      repository: 'https://github.com/react-dropzone/react-dropzone',
    },
    maintainers: [
      {
        email: 'andrey@okonet.ru',
        username: 'okonet',
      },
      {
        email: 'rolandjitsu@gmail.com',
        username: 'rolandjitsu',
      },
      {
        email: 'rmarkins@gmail.com',
        username: 'rxmarbles',
      },
    ],
    name: 'react-dropzone',
    publisher: {
      email: 'rolandjitsu@gmail.com',
      username: 'rolandjitsu',
    },
    scope: 'unscoped',
    version: '11.0.1',
  },
  {
    author: {
      email: 'dan.abramov@me.com',
      name: 'Dan Abramov',
      url: 'https://github.com/gaearon',
    },
    date: '2020-02-18T03:43:52.869Z',
    description: 'Official React bindings for Redux',
    keywords: ['react', 'reactjs', 'redux'],
    links: {
      bugs: 'https://github.com/reduxjs/react-redux/issues',
      homepage: 'https://github.com/reduxjs/react-redux',
      npm: 'https://www.npmjs.com/package/react-redux',
      repository: 'https://github.com/reduxjs/react-redux',
    },
    maintainers: [
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'timdorr@timdorr.com',
        username: 'timdorr',
      },
    ],
    name: 'react-redux',
    publisher: {
      email: 'mark.erikson@gmail.com',
      username: 'acemarke',
    },
    scope: 'unscoped',
    version: '7.2.0',
  },
  {
    author: {
      email: 'case@casesandberg.com',
      name: 'case',
      username: 'case',
    },
    date: '2020-05-04T00:55:19.458Z',
    description:
      'A Collection of Color Pickers from Sketch, Photoshop, Chrome & more',
    keywords: [
      'react',
      'color picker',
      'react-component',
      'colorpicker',
      'picker',
      'sketch',
      'chrome',
      'photoshop',
      'material design',
      'popup',
    ],
    links: {
      homepage: 'http://casesandberg.github.io/react-color/',
      npm: 'https://www.npmjs.com/package/react-color',
    },
    maintainers: [
      {
        email: 'case@casesandberg.com',
        username: 'case',
      },
    ],
    name: 'react-color',
    publisher: {
      email: 'case@casesandberg.com',
      username: 'case',
    },
    scope: 'unscoped',
    version: '2.18.1',
  },
  {
    date: '2020-03-19T19:53:23.004Z',
    description: 'React package for working with the DOM.',
    keywords: ['react'],
    links: {
      bugs: 'https://github.com/facebook/react/issues',
      homepage: 'https://reactjs.org/',
      npm: 'https://www.npmjs.com/package/react-dom',
      repository: 'https://github.com/facebook/react',
    },
    maintainers: [
      {
        email: 'npm@andrewclark.io',
        username: 'acdlite',
      },
      {
        email: 'briandavidvaughn@gmail.com',
        username: 'brianvaughn',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
      {
        email: 'dan.abramov@gmail.com',
        username: 'gaearon',
      },
      {
        email: 'lunaris.ruan@gmail.com',
        username: 'lunaruan',
      },
      {
        email: 'sebastian@calyptus.eu',
        username: 'sebmarkbage',
      },
      {
        email: 'npm@sophiebits.com',
        username: 'sophiebits',
      },
      {
        email: 'dg@domgan.com',
        username: 'trueadm',
      },
      {
        email: 'paul@oshannessy.com',
        username: 'zpao',
      },
    ],
    name: 'react-dom',
    publisher: {
      email: 'npm@andrewclark.io',
      username: 'acdlite',
    },
    scope: 'unscoped',
    version: '16.13.1',
  },
  {
    author: {
      name: 'Felix Kling',
    },
    date: '2020-02-26T00:42:09.515Z',
    description:
      'A CLI and toolkit to extract information from React components for documentation generation.',
    keywords: ['react', 'documentation-generation'],
    links: {
      bugs: 'https://github.com/reactjs/react-docgen/issues',
      homepage: 'https://github.com/reactjs/react-docgen#readme',
      npm: 'https://www.npmjs.com/package/react-docgen',
      repository: 'https://github.com/reactjs/react-docgen',
    },
    maintainers: [
      {
        email: 'daniel@tschinder.de',
        username: 'danez',
      },
      {
        email: 'opensource+npm@fb.com',
        username: 'fb',
      },
    ],
    name: 'react-docgen',
    publisher: {
      email: 'daniel@tschinder.de',
      username: 'danez',
    },
    scope: 'unscoped',
    version: '5.3.0',
  },
  {
    date: '2020-05-15T21:38:04.997Z',
    description:
      'Hooks for building lightweight, fast and extendable datagrids for React',
    keywords: ['react', 'table', 'react-table', 'datagrid'],
    links: {
      bugs: 'https://github.com/tannerlinsley/react-table/issues',
      homepage: 'https://react-table.js.org',
      npm: 'https://www.npmjs.com/package/react-table',
      repository: 'https://github.com/tannerlinsley/react-table',
    },
    maintainers: [
      {
        email: 'tannerlinsley@gmail.com',
        username: 'tannerlinsley',
      },
      {
        email: 'zach.h.sherman@gmail.com',
        username: 'zsherman',
      },
    ],
    name: 'react-table',
    publisher: {
      email: 'tannerlinsley@gmail.com',
      username: 'tannerlinsley',
    },
    scope: 'unscoped',
    version: '7.1.0',
  },
];
