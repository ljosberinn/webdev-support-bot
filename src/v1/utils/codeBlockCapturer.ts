const BACKTICKS = '```';

export type CodeBlockData = {
  code: string;
  language: string;
};

export function createCodeBlockCapturer(langs: string[] = []) {
  const langAlts = langs.join('|');
  return function* (str: string): Generator<CodeBlockData> {
    const langRegex = new RegExp(
      String.raw`${BACKTICKS}(?<language>${langAlts})\n(?<code>[\s\S]+?)\n${BACKTICKS}`,
      'gi'
    );
    const matches = str.matchAll(langRegex);
    for (const {
      groups: { language, code },
    } of matches) {
      yield {
        language: language || '',
        code: code || '',
      };
    }
  };
}
