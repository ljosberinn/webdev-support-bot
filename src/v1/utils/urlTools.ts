import { Message } from 'discord.js';
import { HeadersInit } from 'node-fetch';

import { delayedMessageAutoDeletion } from './delayedMessageAutoDeletion';
import { Provider } from './discordTools';
import { noResults, invalidResponse } from './errors';
import useData from './useData';

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

type ProviderMap = {
  [key in Provider]: {
    search: string;
    color: number;
    createTitle: (term: string) => string;
    icon: string;
    help: string;
    direct?: string;
    getExtendedInfoUrl?: (pkg: string) => string;
  };
};

export const providers: ProviderMap = {
  bundlephobia: {
    color: 0xffffff,
    createTitle: (searchTerm: string) =>
      `Bundlephobia results for *${searchTerm}*`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://bundlephobia.com/api/size?package=${pkg}&record=true`,
    help: '!bundlephobia @chakra-ui/core',
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
  },
  caniuse: {
    color: 0xdb5600,
    createTitle: (searchTerm: string) => `CanIUse results for *${searchTerm}*`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    getExtendedInfoUrl: (text: string) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
    icon: 'https://caniuse.com/img/favicon-128.png',
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
  },
  composer: {
    color: 0xf28d1a,
    createTitle: (searchTerm: string) => `Packagist results for ${searchTerm}`,
    direct: `https://packagist.org/packages/${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://packagist.org/packages/${pkg}.json`,
    help: '!composer sentry',
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
  },
  github: {
    color: 0x24292e,
    createTitle: (searchTerm: string) => `GitHub results for *${searchTerm}*`,
    direct: `https://github.com/${TERM}`,
    help: '!github react',
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
  },
  mdn: {
    color: 0x83d0f2,
    createTitle: (searchTerm: string) => `MDN results for *${searchTerm}*`,
    direct: `https://developer.mozilla.org${TERM}`,
    help: '!mdn localStorage',
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    search: `https://developer.mozilla.org/api/v1/search?q=${SEARCH_TERM}&locale=en-US`,
  },
  npm: {
    color: 0xfb3e44,
    createTitle: (searchTerm: string) => `NPM results for *${searchTerm}*`,
    help: '!npm react',
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
  },
  php: {
    color: 0x8892bf,
    createTitle: (searchTerm: string) => `PHP.net results for *${searchTerm}*`,
    direct: `https://www.php.net/${TERM}`,
    help: '!php echo',
    icon: 'https://www.php.net/images/logos/php-logo.svg',
    search: `https://www.php.net/${SEARCH_TERM}`,
  },
};

export const HELP_KEYWORD = '--help';
export const FORMATTING_KEYWORD = '!formatting';
export const FORMATTING_KEYWORD_ALT = '!format';
export const CODE_KEYWORD = '!code';
export const VSCODE_KEYWORD = '!vscode';
export const JOB_POSTING_KEYWORD = '!post';
export const JQUERY_KEYWORD = '!jquery';
export const POINTS_KEYWORD = '!points';
export const LEADERBOARD_KEYWORD = '!leaderboard';
export const DECAY_KEYWORD = '!decay';
export const MODULE_KEYWORD = '!modules';
export const LOCKFILE_KEYWORD = '!lockfile';
export const FLEXBOX_KEYWORD = '!flexbox';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
export const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|')
  ),
  'i'
);

export const getSearchUrl = (provider: Provider, search: string) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const buildDirectUrl = (provider: Provider, href: string) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const getExtendedInfoUrl = (provider: Provider, term: string) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`
  );
};

interface GetDataParams {
  msg: Message;
  provider: Provider;
  searchTerm: string;
  sanitizeData?: (data: any) => Partial<any>;
  isInvalidData: (data: any) => boolean;
  headers?: HeadersInit;
}

export const getData = async <T>({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}: GetDataParams): Promise<Partial<T>> => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData<T>(searchUrl, 'json', headers);

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};
