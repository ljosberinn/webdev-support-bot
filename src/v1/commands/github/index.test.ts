import { getChosenResult } from '../../utils/discordTools';
import { getData } from '../../utils/urlTools';
import useData from '../../utils/useData';
import { response } from './__fixtures__/response';

import { buildGithubQueryHandler } from '.';

describe('github', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const fetchDetails: jest.MockedFunction<typeof useData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  afterEach(() => jest.clearAllMocks());

  test('returns when request fails', async () => {
    fetch.mockResolvedValue(null);
    const handler = buildGithubQueryHandler(fetch);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      headers: {
        Accept: 'application/vnd.github.v3+json',
      },
      isInvalidData: expect.any(Function),
      msg,
      provider: 'github',
      searchTerm: 'search term',
    });

    expect(choose).not.toBeCalled();
  });

  test('awaits user response when request works', async () => {
    const editMock = jest.fn();
    fetch.mockResolvedValue(response);
    choose.mockResolvedValue({
      created: new Date(),
      description: 'React',
      forks: '',
      issues: 0,
      language: 'en-US',
      name: 'React',
      owner: {
        avatar: 'some_avator',
        name: 'Facebook',
        type: 'something',
      },
      stars: 10000,
      updated: new Date(),
      url: 'https://facebook.com/react',
    });

    fetchDetails.mockResolvedValue({
      error: false,
      json: {
        body: 'MIT',
        conditions: [],
        description: 'MIT',
        featured: true,
        html_url: 'http://choosealicense.com/licenses/mit',
        implementation: 'MIT',
        key: 'key',
        limitations: [],
        name: 'MIT',
        node_id: '',
        permissions: [],
        spdx_id: 'mit',
        url: 'http://choosealicense.com/licenses/mit',
      },
      text: null,
    });

    msg.channel.send.mockResolvedValue({ edit: editMock });

    const handler = buildGithubQueryHandler(fetch, fetchDetails, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      headers: {
        Accept: 'application/vnd.github.v3+json',
      },
      isInvalidData: expect.any(Function),
      msg,
      provider: 'github',
      searchTerm: 'search term',
    });

    expect(msg.channel.send.mock.calls[0]).toMatchSnapshot();
    expect(choose.mock.calls[0]).toMatchSnapshot();
    expect(sendMock.mock.calls[0]).toMatchSnapshot();
  });
});
