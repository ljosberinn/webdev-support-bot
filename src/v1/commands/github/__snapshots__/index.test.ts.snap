// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github awaits user response when request works 1`] = `
Array [
  Object {
    "embed": Object {
      "author": null,
      "color": 2369838,
      "description": "1. [**facebook/react** - *A declarative, efficient, and flexible JavaScript...*](https://github.com/facebook/react)
2. [**duxianwei520/react** - *React+webpack+redux+ant...*](https://github.com/duxianwei520/react)
3. [**discountry/react** - *React docs in Chinese | React 中文文档翻译*](https://github.com/discountry/react)
4. [**Cathy0807/react** - *京东首页构建*](https://github.com/Cathy0807/react)
5. [**react-redux-antd-es6/react** - *基于react的企业后台管理开发框架*](https://github.com/react-redux-antd-es6/react)
6. [**HackYourFuture/React** - *This repository contains all the material for...*](https://github.com/HackYourFuture/React)
7. [**ReactiveCocoa/ReactiveCocoa** - *Cocoa framework and Obj-C dynamism...*](https://github.com/ReactiveCocoa/ReactiveCocoa)
8. [**facebook/react-native** - *A framework for building native apps with...*](https://github.com/facebook/react-native)
9. [**ReactTraining/react-router** - *Declarative routing for React*](https://github.com/ReactTraining/react-router)
10. [**vercel/next.js** - *The React Framework*](https://github.com/vercel/next.js)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
      "fields": Array [],
      "footer": Object {
        "iconURL": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
        "text": "1,394,200 results",
      },
      "title": "GitHub results for *search term*",
      "url": "https://github.com/search?q=search%20term",
    },
  },
]
`;

exports[`github awaits user response when request works 2`] = `
Array [
  Object {
    "edit": [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "embed": Object {
              "author": Object {
                "icon_url": "some_avator",
                "name": "Facebook ",
                "url": "https://github.com/Facebook",
              },
              "color": 2369838,
              "description": "React",
              "fields": Array [
                Object {
                  "inline": false,
                  "name": "clone via...",
                  "value": "\`\`\`bash
git clone https://facebook.com/react
\`\`\`",
                },
                Object {
                  "inline": true,
                  "name": ":warning: open issues",
                  "value": "[0](https://facebook.com/react/issues)",
                },
                Object {
                  "inline": true,
                  "name": ":star: stars",
                  "value": "[10,000](https://facebook.com/react/stargazers)",
                },
                Object {
                  "inline": true,
                  "name": ":fork_and_knife: forks",
                  "value": "[](https://facebook.com/react/network/members)",
                },
                Object {
                  "inline": false,
                  "name": ":writing_hand: language",
                  "value": "en-US",
                },
              ],
              "footer": Object {
                "iconURL": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
                "text": "updated less than a minute ago - created less than a minute ago",
              },
              "title": "React",
              "url": "https://facebook.com/react",
            },
          },
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  Object {
    "channel": Object {
      "send": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "embed": Object {
                "author": null,
                "color": 2369838,
                "description": "1. [**facebook/react** - *A declarative, efficient, and flexible JavaScript...*](https://github.com/facebook/react)
2. [**duxianwei520/react** - *React+webpack+redux+ant...*](https://github.com/duxianwei520/react)
3. [**discountry/react** - *React docs in Chinese | React 中文文档翻译*](https://github.com/discountry/react)
4. [**Cathy0807/react** - *京东首页构建*](https://github.com/Cathy0807/react)
5. [**react-redux-antd-es6/react** - *基于react的企业后台管理开发框架*](https://github.com/react-redux-antd-es6/react)
6. [**HackYourFuture/React** - *This repository contains all the material for...*](https://github.com/HackYourFuture/React)
7. [**ReactiveCocoa/ReactiveCocoa** - *Cocoa framework and Obj-C dynamism...*](https://github.com/ReactiveCocoa/ReactiveCocoa)
8. [**facebook/react-native** - *A framework for building native apps with...*](https://github.com/facebook/react-native)
9. [**ReactTraining/react-router** - *Declarative routing for React*](https://github.com/ReactTraining/react-router)
10. [**vercel/next.js** - *The React Framework*](https://github.com/vercel/next.js)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
                "fields": Array [],
                "footer": Object {
                  "iconURL": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
                  "text": "1,394,200 results",
                },
                "title": "GitHub results for *search term*",
                "url": "https://github.com/search?q=search%20term",
              },
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "delete": [MockFunction],
    "reply": [MockFunction],
  },
  Array [
    Object {
      "created": "2013-05-24T16:15:54Z",
      "description": "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
      "forks": "29165",
      "homepage": "https://reactjs.org",
      "issues": 613,
      "language": "JavaScript",
      "license": Object {
        "name": "MIT License",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "facebook/react",
      "owner": Object {
        "avatar": "https://avatars3.githubusercontent.com/u/69631?v=4",
        "name": "facebook",
        "type": "Organization",
      },
      "stars": 149876,
      "updated": "2020-06-04T11:35:07Z",
      "url": "https://github.com/facebook/react",
    },
    Object {
      "created": "2016-12-02T13:08:43Z",
      "description": " React+webpack+redux+ant design+axios+less全家桶后台管理框架",
      "forks": "1317",
      "homepage": "",
      "issues": 0,
      "language": "JavaScript",
      "license": Object {
        "name": "MIT License",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "duxianwei520/react",
      "owner": Object {
        "avatar": "https://avatars1.githubusercontent.com/u/3249653?v=4",
        "name": "duxianwei520",
        "type": "User",
      },
      "stars": 3649,
      "updated": "2020-06-04T08:25:48Z",
      "url": "https://github.com/duxianwei520/react",
    },
    Object {
      "created": "2017-05-09T15:09:00Z",
      "description": "React docs in Chinese | React 中文文档翻译",
      "forks": "538",
      "homepage": "https://react.docschina.org/",
      "issues": 14,
      "language": "JavaScript",
      "license": Object {
        "name": "Creative Commons Attribution 4.0 International",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "discountry/react",
      "owner": Object {
        "avatar": "https://avatars0.githubusercontent.com/u/4507101?v=4",
        "name": "discountry",
        "type": "User",
      },
      "stars": 941,
      "updated": "2020-06-03T07:08:04Z",
      "url": "https://github.com/discountry/react",
    },
    Object {
      "created": "2016-11-02T10:18:45Z",
      "description": "京东首页构建",
      "forks": "366",
      "homepage": null,
      "issues": 0,
      "language": "JavaScript",
      "license": undefined,
      "name": "Cathy0807/react",
      "owner": Object {
        "avatar": "https://avatars0.githubusercontent.com/u/20653643?v=4",
        "name": "Cathy0807",
        "type": "User",
      },
      "stars": 645,
      "updated": "2020-05-29T09:54:23Z",
      "url": "https://github.com/Cathy0807/react",
    },
    Object {
      "created": "2016-12-28T07:44:04Z",
      "description": "基于react的企业后台管理开发框架",
      "forks": "323",
      "homepage": "",
      "issues": 7,
      "language": "JavaScript",
      "license": Object {
        "name": "MIT License",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "react-redux-antd-es6/react",
      "owner": Object {
        "avatar": "https://avatars3.githubusercontent.com/u/24805142?v=4",
        "name": "react-redux-antd-es6",
        "type": "Organization",
      },
      "stars": 708,
      "updated": "2020-06-02T01:42:05Z",
      "url": "https://github.com/react-redux-antd-es6/react",
    },
    Object {
      "created": "2017-06-06T09:58:23Z",
      "description": "This repository contains all the material for the HackYourFuture module \\"React.js: Building dynamic UIs with modern JavaScript\\"",
      "forks": "324",
      "homepage": "",
      "issues": 9,
      "language": null,
      "license": undefined,
      "name": "HackYourFuture/React",
      "owner": Object {
        "avatar": "https://avatars2.githubusercontent.com/u/20858568?v=4",
        "name": "HackYourFuture",
        "type": "Organization",
      },
      "stars": 78,
      "updated": "2020-06-02T12:00:55Z",
      "url": "https://github.com/HackYourFuture/React",
    },
    Object {
      "created": "2012-03-02T22:11:24Z",
      "description": "Cocoa framework and Obj-C dynamism bindings for ReactiveSwift.",
      "forks": "3551",
      "homepage": "",
      "issues": 48,
      "language": "Swift",
      "license": undefined,
      "name": "ReactiveCocoa/ReactiveCocoa",
      "owner": Object {
        "avatar": "https://avatars0.githubusercontent.com/u/3422977?v=4",
        "name": "ReactiveCocoa",
        "type": "Organization",
      },
      "stars": 19811,
      "updated": "2020-06-03T11:29:10Z",
      "url": "https://github.com/ReactiveCocoa/ReactiveCocoa",
    },
    Object {
      "created": "2015-01-09T18:10:16Z",
      "description": "A framework for building native apps with React.",
      "forks": "19491",
      "homepage": "https://reactnative.dev",
      "issues": 914,
      "language": "JavaScript",
      "license": undefined,
      "name": "facebook/react-native",
      "owner": Object {
        "avatar": "https://avatars3.githubusercontent.com/u/69631?v=4",
        "name": "facebook",
        "type": "Organization",
      },
      "stars": 87733,
      "updated": "2020-06-04T11:29:30Z",
      "url": "https://github.com/facebook/react-native",
    },
    Object {
      "created": "2014-05-16T22:22:51Z",
      "description": "Declarative routing for React",
      "forks": "8153",
      "homepage": "https://reacttraining.com/react-router/",
      "issues": 56,
      "language": "JavaScript",
      "license": Object {
        "name": "MIT License",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "ReactTraining/react-router",
      "owner": Object {
        "avatar": "https://avatars1.githubusercontent.com/u/11823761?v=4",
        "name": "ReactTraining",
        "type": "Organization",
      },
      "stars": 40730,
      "updated": "2020-06-04T09:49:03Z",
      "url": "https://github.com/ReactTraining/react-router",
    },
    Object {
      "created": "2016-10-05T23:32:51Z",
      "description": "The React Framework",
      "forks": "7449",
      "homepage": "https://nextjs.org",
      "issues": 494,
      "language": "JavaScript",
      "license": Object {
        "name": "MIT License",
        "url": "http://choosealicense.com/licenses/mit",
      },
      "name": "vercel/next.js",
      "owner": Object {
        "avatar": "https://avatars0.githubusercontent.com/u/14985020?v=4",
        "name": "vercel",
        "type": "Organization",
      },
      "stars": 49194,
      "updated": "2020-06-04T11:07:19Z",
      "url": "https://github.com/vercel/next.js",
    },
  ],
]
`;

exports[`github awaits user response when request works 3`] = `
Array [
  Object {
    "embed": Object {
      "author": null,
      "color": 2369838,
      "description": "1. [**facebook/react** - *A declarative, efficient, and flexible JavaScript...*](https://github.com/facebook/react)
2. [**duxianwei520/react** - *React+webpack+redux+ant...*](https://github.com/duxianwei520/react)
3. [**discountry/react** - *React docs in Chinese | React 中文文档翻译*](https://github.com/discountry/react)
4. [**Cathy0807/react** - *京东首页构建*](https://github.com/Cathy0807/react)
5. [**react-redux-antd-es6/react** - *基于react的企业后台管理开发框架*](https://github.com/react-redux-antd-es6/react)
6. [**HackYourFuture/React** - *This repository contains all the material for...*](https://github.com/HackYourFuture/React)
7. [**ReactiveCocoa/ReactiveCocoa** - *Cocoa framework and Obj-C dynamism...*](https://github.com/ReactiveCocoa/ReactiveCocoa)
8. [**facebook/react-native** - *A framework for building native apps with...*](https://github.com/facebook/react-native)
9. [**ReactTraining/react-router** - *Declarative routing for React*](https://github.com/ReactTraining/react-router)
10. [**vercel/next.js** - *The React Framework*](https://github.com/vercel/next.js)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
      "fields": Array [],
      "footer": Object {
        "iconURL": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
        "text": "1,394,200 results",
      },
      "title": "GitHub results for *search term*",
      "url": "https://github.com/search?q=search%20term",
    },
  },
]
`;
